.github/actions/environment/action.yml

```diff
@@ -8,4 +8,13 @@ runs:
         echo TEST_TIMEOUT=310 | tee -a $GITHUB_ENV
         echo NO_OUTPUT_TIMEOUT=320 | tee -a $GITHUB_ENV
         echo PRESERVE_ENVVARS=REPLICATION_SYNC_TIMEOUT,TEST_TIMEOUT,NO_OUTPUT_TIMEOUT | tee -a $GITHUB_ENV
+        # This switching off swap command will not work as github step
+        # run from inside github 'container' tag. Also it will fail to
+        # run from it. So running it only outside of docker container.
+        # NOTE: To switch off swap from inside github 'container' tag
+        # additional memory flags should be added to its 'options' tag:
+        #   options: '--memory=<some value, like 7G> --memory-swap=<the same value as for memory option>'
+        if which docker && sudo ls /sbin/swapoff ; then
```

Comment:
I would use something like [this](https://stackoverflow.com/a/20012536/1598057) to determine whether we're in the docker container (it is used in c9802bd82b4893b57a068d20b79dff5304e853a7).

Reply:
Right, I saw this way and even tried it. But after thinking about it, decided that this way is not good for us, because it uses the check on paths, which not guaranteed to be the same forever. Also I found in mentioned discussion that people already tried to use configuration files that changed between docker's versions and broke their tries. Checking any of the 'docker' path entry is a bad way, also some files like this may have 'init' and the other 'systemd' and it makes the check too huge. Also we don't need to check LXC.

In my way I check docker availability which always exists due to Github Actions use containers and won't exists in docker containers, because we don't specially need it and to have it we'll need to do special images with it.

Anyway if we really against of my easy way let's use the way that have real evidences of searching values. Due to init 1 process can be only 'init' either 'systemd' (may be in future it will be some other name, but not now) outside of container:
```
cat /proc/1/sched | head -n1 | grep "[^init |^systemd ]"
```

Reply:
I've changed my initial variant to the written above.

Reply:
I don't strongly insist on any of those ways. My thoughts are the following.

Our task is to make a simple (but reliable) check whether we're inside a docker container. There are docker-in-docker images and so `which docker` (`type docker`, `command -v docker`) may fail if we'll use such image and will forget to update the check. The dind images are not the usual thing, but more like a hack, so unlikely we'll be hit by the issue.

Are there other problems with this approach? Okay, maybe, if we'll have a custom tiny runner (arm64?) without docker, but (surprisingly) with swap enabled... Small chance, but possible.

What about `/proc/1/cgroup`? To be honest, I don't get your answer. When the check fails? I tried it on my host system, in `docker run -it centos:7` and `docker run --init -it centos:7` and it outputs paths with `/docker/` only when run inside docker.

Regarding LXC: I guess we can assume we're on the new docker, which uses `/docker/` in paths.

Regarding `/proc/1/sched`: I have no such file. My kernel (gentoo-sources-5.4.10) have `CONFIG_SCHED_DEBUG` disabled, however even if it would be enabled: the help states that it enables `/proc/sched_debug`. May it is something relevant only to older kernel versions?

Another idea comes into the mind: can we just call `swapoff -a`, without any host/docker checks? Are there any negative consequences? As I see, it returns zero if there is nothing to disable.

Reply:
> Another idea comes into the mind: can we just call `swapoff -a`, without any host/docker checks? Are there any negative consequences? As I see, it returns zero if there is nothing to disable.

My bad, I tried to verify it with commented swap partition in /etc/fstab: so there was nothing to enable / disable. Actually it gives 1 exit code and 'swapoff: Not superuser.' error message from a docker container.

Reply:
> I've changed my initial variant to the written above.

Missed the comment, sorry. This check does not work on my host (see above). Maybe it is not the best option. I'll reopen the discussion to continue it.

Reply:
Hm. Another thing: AFAIR, the action is run *outside* of the container even in docker jobs? So, any check whether we're inside a docker container, are not sufficient.

How about just invoke `sudo swapoff -a || true` and than check `[ -e /proc/meminfo ] && [ "$(grep SwapTotal: /proc/meminfo | awk '{ print $2; }')" = "0" ]`? It looks straightforward. The `/proc/meminfo` gives a host information even inside the container, but it fit our needs.

Reply:
Right, your suggestion is completely informative, so I've used it instead of mine variant.

Reply:
> Hm. Another thing: AFAIR, the action is run outside of the container even in docker jobs? So, any check whether we're inside a docker container, are not sufficient.

For the sake of eventual correctness. I verified, it is not so. If we'll run an action inside a container job, it'll run inside the container, as expected.

My experiment:

.github/actions/am_i_in_docker/action.yml:

```yaml
name: Am I in Docker?
runs:
  using: composite
  steps:
    - run: cat /proc/1/cgroup
      shell: bash
```

.github/workflows/hello_world.yml:

```yaml
name: Hello, world!

on: [push]

jobs:
  native_job:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - uses: ./.github/actions/am_i_in_docker

  container_job:
    runs-on: ubuntu-latest
    container: centos:7
    steps:
      - uses: actions/checkout@v2
      - uses: ./.github/actions/am_i_in_docker
```

Results:

native_job:

```
12:blkio:/init.scope
11:perf_event:/
10:memory:/init.scope
9:devices:/init.scope
8:cpuset:/
7:hugetlb:/
6:rdma:/
5:cpu,cpuacct:/init.scope
4:pids:/init.scope
3:net_cls,net_prio:/
2:freezer:/
1:name=systemd:/init.scope
0::/init.scope
```

container_job:

```
12:rdma:/
11:devices:/actions_job/593dfe2b03cf722f3da6fbdb3d523303b2306ced2f52fd53e50c34b727f677cb
10:hugetlb:/actions_job/593dfe2b03cf722f3da6fbdb3d523303b2306ced2f52fd53e50c34b727f677cb
9:memory:/actions_job/593dfe2b03cf722f3da6fbdb3d523303b2306ced2f52fd53e50c34b727f677cb
8:cpuset:/actions_job/593dfe2b03cf722f3da6fbdb3d523303b2306ced2f52fd53e50c34b727f677cb
7:pids:/actions_job/593dfe2b03cf722f3da6fbdb3d523303b2306ced2f52fd53e50c34b727f677cb
6:blkio:/actions_job/593dfe2b03cf722f3da6fbdb3d523303b2306ced2f52fd53e50c34b727f677cb
5:freezer:/actions_job/593dfe2b03cf722f3da6fbdb3d523303b2306ced2f52fd53e50c34b727f677cb
4:net_cls,net_prio:/actions_job/593dfe2b03cf722f3da6fbdb3d523303b2306ced2f52fd53e50c34b727f677cb
3:cpu,cpuacct:/actions_job/593dfe2b03cf722f3da6fbdb3d523303b2306ced2f52fd53e50c34b727f677cb
2:perf_event:/actions_job/593dfe2b03cf722f3da6fbdb3d523303b2306ced2f52fd53e50c34b727f677cb
1:name=systemd:/actions_job/593dfe2b03cf722f3da6fbdb3d523303b2306ced2f52fd53e50c34b727f677cb
0::/system.slice/containerd.service
```